From 23e6a324b30c774e0049192eec5398181551f34b Mon Sep 17 00:00:00 2001
From: Heyu Hu <huheyu2@huawei.com>
Date: Thu, 12 Feb 2026 20:45:46 +0800
Subject: [PATCH] urma: implement ipourma network device operations

Offering: URMA
URMA inclusion
category: feature
bugzilla: NA

----------------------------------------------

This patch implement ipourma's netdev_ops,ethtool_ops and
netlink_ops, providing the capability to read and configure
fundamental network device properties, as well as enabling
data plane communication.

Signed-off-by: Heyu Hu <huheyu2@huawei.com>
---
 .../ub/urma/ulp/ipourma/ipourma_addr_res.c    |  29 +-
 .../ub/urma/ulp/ipourma/ipourma_addr_res.h    |   6 +-
 drivers/ub/urma/ulp/ipourma/ipourma_ethtool.c | 183 +++++++++++-
 drivers/ub/urma/ulp/ipourma/ipourma_netdev.c  | 178 ++++++++++-
 drivers/ub/urma/ulp/ipourma/ipourma_netlink.c |  94 +++++-
 drivers/ub/urma/ulp/ipourma/ipourma_res.c     | 224 ++++++++++++++
 drivers/ub/urma/ulp/ipourma/ipourma_res.h     |   3 +
 drivers/ub/urma/ulp/ipourma/ipourma_types.h   |   3 +-
 drivers/ub/urma/ulp/ipourma/ipourma_ub.c      | 280 +++++++++++++++++-
 drivers/ub/urma/ulp/ipourma/ipourma_ub.h      |   8 +
 10 files changed, 990 insertions(+), 18 deletions(-)

diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_addr_res.c b/drivers/ub/urma/ulp/ipourma/ipourma_addr_res.c
index 88368aca22e9..54b4d5befa8e 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_addr_res.c
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_addr_res.c
@@ -10,6 +10,31 @@
 #include "ipourma_err.h"
 #include "ipourma_addr_res.h"
 
+STATIC void ipourma_resolve_eids_from_ipv6(struct net_device *dev,
+	struct in6_addr *src, struct in6_addr *dst,
+	union ubcore_eid *src_eid, union ubcore_eid *dst_eid)
+{
+	memcpy(src_eid, src, UBCORE_EID_SIZE);
+	memcpy(dst_eid, dst, UBCORE_EID_SIZE);
+}
+
+void ipourma_resolve_eids(struct net_device *dev, struct sk_buff *skb, u16 proto,
+			  union ubcore_eid *src_eid, union ubcore_eid *dst_eid)
+{
+	struct ipv6hdr *ipv6h;
+
+	switch (proto) {
+	case ETH_P_IPV6:
+		ipv6h = (struct ipv6hdr *)skb_network_header(skb);
+		ipourma_resolve_eids_from_ipv6(dev, &ipv6h->saddr,
+						&ipv6h->daddr, src_eid, dst_eid);
+		break;
+	default:
+		netdev_dbg(dev, "%s\n", ipourma_err_desc(IPOURMA_UNSUPPORTED_ETH_PROTO));
+		break;
+	}
+}
+
 STATIC int ipourma_resolve_ipv6_from_eids(struct net_device *dev,
 	union ubcore_eid *eid, struct in6_addr *addr)
 {
@@ -17,8 +42,8 @@ STATIC int ipourma_resolve_ipv6_from_eids(struct net_device *dev,
 	return IPOURMA_OK;
 }
 
-int ipourma_resolve_ipaddr(struct net_device *dev,
-	u16 proto, union ubcore_eid *eid, struct in6_addr *addr)
+int ipourma_resolve_ipaddr(struct net_device *dev, u16 proto,
+			   union ubcore_eid *eid, struct in6_addr *addr)
 {
 	switch (proto) {
 	case ETH_P_IPV6:
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_addr_res.h b/drivers/ub/urma/ulp/ipourma/ipourma_addr_res.h
index 94dc2803a7d6..a7772c2fdcac 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_addr_res.h
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_addr_res.h
@@ -10,7 +10,9 @@
 
 #include "ipourma_types.h"
 
-int ipourma_resolve_ipaddr(struct net_device *dev,
-	u16 proto, union ubcore_eid *eid, struct in6_addr *addr);
+void ipourma_resolve_eids(struct net_device *dev, struct sk_buff *skb, u16 proto,
+			  union ubcore_eid *src_eid, union ubcore_eid *dst_eid);
+int ipourma_resolve_ipaddr(struct net_device *dev, u16 proto,
+			   union ubcore_eid *eid, struct in6_addr *addr);
 
 #endif
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_ethtool.c b/drivers/ub/urma/ulp/ipourma/ipourma_ethtool.c
index 1943ad56c5b6..89bb4ef5cfda 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_ethtool.c
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_ethtool.c
@@ -16,7 +16,188 @@
 #define IPOURMA_DRIVER_VERSION	"0.1.0"
 #define IPOURMA_FW_VERSION	"0.1.0"
 
+struct ipourma_stats {
+	char stat_string[ETH_GSTRING_LEN];
+	int stat_offset;
+};
+
+#define IPOURMA_NETDEV_STAT(m) { \
+		.stat_string = #m, \
+		.stat_offset = offsetof(struct rtnl_link_stats64, m) }
+
+static const struct ipourma_stats ipourma_gstrings_stats[] = {
+	IPOURMA_NETDEV_STAT(rx_packets),
+	IPOURMA_NETDEV_STAT(tx_packets),
+	IPOURMA_NETDEV_STAT(rx_bytes),
+	IPOURMA_NETDEV_STAT(tx_bytes),
+	IPOURMA_NETDEV_STAT(tx_errors),
+	IPOURMA_NETDEV_STAT(rx_dropped),
+	IPOURMA_NETDEV_STAT(tx_dropped),
+};
+
+#define IPOURMA_GLOBAL_STATS_LEN	ARRAY_SIZE(ipourma_gstrings_stats)
+
+STATIC int ub_speed_enum_to_int(enum ubcore_speed speed)
+{
+	switch (speed) {
+	case UBCORE_SP_10M:
+		return SPEED_10;
+	case UBCORE_SP_100M:
+		return SPEED_100;
+	case UBCORE_SP_1G:
+		return SPEED_1000;
+	case UBCORE_SP_2_5G:
+		return SPEED_2500;
+	case UBCORE_SP_5G:
+		return SPEED_5000;
+	case UBCORE_SP_10G:
+		return SPEED_10000;
+	case UBCORE_SP_14G:
+		return SPEED_14000;
+	case UBCORE_SP_25G:
+		return SPEED_25000;
+	case UBCORE_SP_40G:
+		return SPEED_40000;
+	case UBCORE_SP_50G:
+		return SPEED_50000;
+	case UBCORE_SP_100G:
+		return SPEED_100000;
+	case UBCORE_SP_200G:
+		return SPEED_200000;
+	case UBCORE_SP_400G:
+		return SPEED_400000;
+	case UBCORE_SP_800G:
+		return SPEED_800000;
+	}
+
+	return SPEED_UNKNOWN;
+}
+
+STATIC int ipourma_get_link_ksettings(struct net_device *netdev,
+					struct ethtool_link_ksettings *cmd)
+{
+	struct ipourma_dev_priv *priv;
+	struct ubcore_device *dev;
+	struct ubcore_device_status status;
+	int ret, speed;
+
+	if (IS_ERR_OR_NULL(netdev) || IS_ERR_OR_NULL(cmd))
+		return -EINVAL;
+	priv = netdev_priv(netdev);
+	dev = priv->urma_dev;
+
+	if (!netif_carrier_ok(netdev)) {
+		cmd->base.speed = SPEED_UNKNOWN;
+		cmd->base.duplex = DUPLEX_UNKNOWN;
+		return 0;
+	}
+
+	if (IS_ERR_OR_NULL(priv->urma_dev->ops) ||
+		IS_ERR_OR_NULL(priv->urma_dev->ops->query_device_status))
+		return -EINVAL;
+
+	ret = priv->urma_dev->ops->query_device_status(dev, &status);
+	if (ret)
+		return ret;
+
+	/* For IP over URMA, each port corresponds to a ubcore_device */
+	speed = ub_speed_enum_to_int(status.port_status[0].active_speed);
+	if (speed < 0)
+		return -EINVAL;
+
+	cmd->base.speed		 = (uint32_t)speed;
+	cmd->base.duplex	 = DUPLEX_FULL;
+	cmd->base.phy_address	 = 0xFF;
+	cmd->base.autoneg	 = AUTONEG_DISABLE;
+	cmd->base.port		 = PORT_FIBRE;
+
+	return 0;
+}
+
+STATIC void ipourma_get_drvinfo(struct net_device *netdev,
+			      struct ethtool_drvinfo *drvinfo)
+{
+	struct ipourma_dev_priv *priv;
+
+	if (IS_ERR_OR_NULL(netdev) || IS_ERR_OR_NULL(drvinfo))
+		return;
+
+	priv = netdev_priv(netdev);
+
+	strscpy(drvinfo->version, IPOURMA_DRIVER_VERSION,
+		sizeof(drvinfo->version));
+	strscpy(drvinfo->fw_version, IPOURMA_FW_VERSION,
+		sizeof(drvinfo->fw_version));
+
+	if (priv->urma_dev->dev.parent) {
+		strscpy(drvinfo->bus_info, dev_name(priv->urma_dev->dev.parent),
+			sizeof(drvinfo->bus_info));
+	}
+
+	strscpy(drvinfo->driver, "ub_ipourma", sizeof(drvinfo->driver));
+}
+
+STATIC void ipourma_get_strings(struct net_device __always_unused *dev,
+			      u32 stringset, u8 *data)
+{
+	u8 *p = data;
+	int i;
+
+	if (IS_ERR_OR_NULL(dev) || IS_ERR_OR_NULL(data))
+		return;
+	switch (stringset) {
+	case ETH_SS_STATS:
+		for (i = 0; i < IPOURMA_GLOBAL_STATS_LEN; i++) {
+			memcpy(p, ipourma_gstrings_stats[i].stat_string,
+				ETH_GSTRING_LEN);
+			p += ETH_GSTRING_LEN;
+		}
+		break;
+	default:
+		break;
+	}
+}
+
+STATIC void ipourma_get_ethtool_stats(struct net_device *dev,
+				    struct ethtool_stats __always_unused *stats,
+				    u64 *data)
+{
+	int i;
+	struct net_device_stats *net_stats;
+	u8 *p;
+
+	if (IS_ERR_OR_NULL(dev) || IS_ERR_OR_NULL(data))
+		return;
+	net_stats = &dev->stats;
+	p = (u8 *)net_stats;
+	for (i = 0; i < IPOURMA_GLOBAL_STATS_LEN; i++)
+		data[i] = *(u64 *)(p + ipourma_gstrings_stats[i].stat_offset);
+}
+
+STATIC int ipourma_get_sset_count(struct net_device __always_unused *dev,
+				 int sset)
+{
+	if (IS_ERR_OR_NULL(dev))
+		return -EINVAL;
+	switch (sset) {
+	case ETH_SS_STATS:
+		return IPOURMA_GLOBAL_STATS_LEN;
+	default:
+		break;
+	}
+	return -EOPNOTSUPP;
+}
+
+static const struct ethtool_ops ipourma_ethtool_ops = {
+	.get_link_ksettings	= ipourma_get_link_ksettings,
+	.get_drvinfo		= ipourma_get_drvinfo,
+	.get_strings		= ipourma_get_strings,
+	.get_ethtool_stats	= ipourma_get_ethtool_stats,
+	.get_sset_count		= ipourma_get_sset_count,
+	.get_link			= ethtool_op_get_link,
+};
+
 void ipourma_set_ethtool_ops(struct net_device *dev)
 {
-	// this function will be filled in the next commit
+	dev->ethtool_ops = &ipourma_ethtool_ops;
 }
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_netdev.c b/drivers/ub/urma/ulp/ipourma/ipourma_netdev.c
index 9500c8a85ee5..a5553a078ade 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_netdev.c
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_netdev.c
@@ -24,6 +24,8 @@
 #include "ipourma_ub.h"
 #include "ipourma_main.h"
 #include "ipourma_ip.h"
+#include "ipourma_err.h"
+#include "ipourma_utils.h"
 #include "ipourma_addr_res.h"
 #include "ipourma_ethtool.h"
 #include "ipourma_netlink.h"
@@ -161,7 +163,15 @@ STATIC void ipourma_add_route_entry(struct work_struct *work)
 
 STATIC int ipourma_ndo_init(struct net_device *dev)
 {
-	// this function will be filled in the next commit
+	if (IS_ERR_OR_NULL(dev))
+		return -EINVAL;
+	/* supports scatter/gather */
+	dev->features |= NETIF_F_SG;
+	/* HW DO NOT support TSO */
+	dev->hw_features &= ~NETIF_F_FRAGLIST;
+
+	ipourma_urma_dev_init(dev);
+
 	return 0;
 }
 
@@ -170,45 +180,185 @@ STATIC void ipourma_ndo_uninit(struct net_device *dev)
 	// do uninit in ipourma_stop
 }
 
+STATIC inline void ipourma_napi_enable(struct net_device *dev)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+
+	napi_enable(&priv->napi_send);
+	napi_enable(&priv->napi_recv);
+}
+
+STATIC inline void ipourma_rearm_jfc(struct ipourma_dev_priv *priv)
+{
+	if (IS_ERR_OR_NULL(priv->rx_jfc) || ubcore_rearm_jfc(priv->rx_jfc, false) != 0) {
+		netdev_dbg(priv->dev, "%s\n", ipourma_err_desc(IPOURMA_REARM_JFC_FAILED));
+		priv->runtime_stats.rx_stats.rearm_failed++;
+	} else
+		priv->runtime_stats.rx_stats.rearm_success++;
+	if (IS_ERR_OR_NULL(priv->tx_jfc) || ubcore_rearm_jfc(priv->tx_jfc, false) != 0) {
+		netdev_dbg(priv->dev, "%s\n", ipourma_err_desc(IPOURMA_REARM_JFC_FAILED));
+		priv->runtime_stats.tx_stats.rearm_failed++;
+	} else
+		priv->runtime_stats.tx_stats.rearm_success++;
+}
+
 STATIC int ipourma_open(struct net_device *dev)
 {
-	// this function will be filled in the next commit
+	struct ipourma_dev_priv *priv;
+
+	if (IS_ERR_OR_NULL(dev))
+		return -EINVAL;
+	priv = netdev_priv(dev);
+
+	atomic_set(&(priv->need_set_ip_route), 0);
+	if (IS_ERR_OR_NULL(priv->net_config_wq))
+		return -EINVAL;
+	queue_work(priv->net_config_wq, &(priv->set_ip));
+	queue_work(priv->net_config_wq, &(priv->set_route));
+
+	ipourma_restart_rings(priv);
+	ipourma_napi_enable(dev);
+	netif_start_queue(dev);
+	dev->flags |= IFF_RUNNING;
+	netif_carrier_on(dev);
+	set_bit(IPOURMA_DEV_ADMIN_UP, &priv->flags);
+	ipourma_rearm_jfc(priv);
+	netdev_info(dev, "Device opened.\n");
+
 	return 0;
 }
 
 STATIC inline void ipourma_napi_disable(struct net_device *dev)
 {
-	// this function will be filled in the next commit
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+
+	napi_disable(&priv->napi_send);
+	napi_disable(&priv->napi_recv);
 }
 
 STATIC int ipourma_stop(struct net_device *dev)
 {
-	// this function will be filled in the next commit
+	struct ipourma_dev_priv *priv;
+
+	if (IS_ERR_OR_NULL(dev))
+		return -EINVAL;
+	priv = netdev_priv(dev);
+
+	ipourma_napi_disable(dev);
+	clear_bit(IPOURMA_DEV_ADMIN_UP, &priv->flags);
+	netif_stop_queue(dev);
+	if (!IS_ERR_OR_NULL(priv->rx_wq))
+		flush_workqueue(priv->rx_wq);
+	ipourma_reset_rings(priv);
+	ipourma_lru_clear(&priv->tjetty_lru);
+	netif_carrier_off(dev);
+	dev->flags &= ~IFF_RUNNING;
+	queue_work(priv->net_config_wq, &(priv->unset_route));
+
+	netdev_info(dev, "Device closed.\n");
+
 	return 0;
 }
 
 STATIC int ipourma_change_mtu(struct net_device *dev, int mtu)
 {
-	// this function will be filled in the next commit
+	bool flag;
+
+	if (IS_ERR_OR_NULL(dev))
+		return -EINVAL;
+
+	/* if device is running, then user should not change the mtu */
+	if (netif_running(dev)) {
+		pr_err("cannot change mtu when device is running\n");
+		return -EINVAL;
+	}
+
+	/* check ranges */
+	if ((mtu < IPOURMA_MIN_MTU) || (mtu > IPOURMA_MAX_MTU))
+		return -EINVAL;
+
+	flag = netif_carrier_ok(dev);
+	netif_carrier_off(dev);
+	dev->mtu = (uint32_t)mtu;
+	if (flag)
+		netif_carrier_on(dev);
+	pr_info("change mtu to %d success!\n", dev->mtu);
 	return 0;
 }
 
 STATIC netdev_features_t ipourma_fix_features(struct net_device *dev,
 	netdev_features_t features)
 {
-	// this function will be filled in the next commit
+	if (IS_ERR_OR_NULL(dev))
+		return -EINVAL;
+
+	features &= ~(NETIF_F_GSO | NETIF_F_TSO | NETIF_F_SG);
 	return features;
 }
 
 STATIC netdev_tx_t ipourma_start_xmit(struct sk_buff *skb, struct net_device *dev)
 {
-	// this function will be filled in the next commit
+	union ubcore_eid src_eid, dst_eid;
+	struct ipourma_dev_priv *priv;
+	u16 proto;
+	int ret;
+
+	if (IS_ERR_OR_NULL(dev) || IS_ERR_OR_NULL(skb) ||
+		IS_ERR_OR_NULL(skb->head) || IS_ERR_OR_NULL(skb->data))
+		return -EINVAL;
+
+	priv  = netdev_priv(dev);
+	priv->runtime_stats.tx_stats.num_recv_pkts_from_kernel++;
+	proto = ntohs(skb->protocol);
+
+	pr_skb_head_plus_linear(skb, "start xmit");
+	/* only support IPv6 */
+	if (proto != ETH_P_IPV6) {
+		priv->runtime_stats.tx_stats.not_ipv6_proto++;
+		netdev_dbg(dev, "Unsupported ether type: %u", proto);
+		goto drop_out;
+	}
+	/* check skb */
+	ret = ipourma_check_skb(dev, skb);
+	if (ret != IPOURMA_OK)
+		goto drop_out;
+
+	ipourma_resolve_eids(dev, skb, proto, &src_eid, &dst_eid);
+
+	ret = ipourma_xmit(dev, skb, &src_eid, &dst_eid);
+	if (ret != IPOURMA_OK)
+		goto err_out;
+
+	return NETDEV_TX_OK;
+
+drop_out:
+	dev->stats.tx_dropped++;
+	dev_kfree_skb_any(skb);
+	return NETDEV_TX_OK;
+err_out:
+	dev->stats.tx_errors++;
+	dev_kfree_skb_any(skb);
 	return NETDEV_TX_OK;
 }
 
 STATIC void ipourma_timeout(struct net_device *dev, unsigned int txqueue)
 {
-	// this function will be filled in the next commit
+	struct ipourma_dev_priv *priv;
+
+	if (IS_ERR_OR_NULL(dev))
+		return;
+
+	priv = netdev_priv(dev);
+	netdev_err(dev, "Transmit timeout at %ld, latency %d\n",
+		jiffies, jiffies_to_msecs(jiffies - dev_trans_start(dev)));
+	if (IS_ERR_OR_NULL(priv->eid_info))
+		return;
+	for (int i = 0; i < IPOURMA_MAX_EID_CNT; i++) {
+		if (eid_is_empty(&priv->eid_info[i].eid))
+			continue;
+		netdev_err(dev, "queue stopped %d, tx_head %u, tx_tail %u\n",
+			netif_queue_stopped(dev), priv->tx_head[i], priv->tx_tail[i]);
+	}
 }
 
 /*
@@ -221,7 +371,17 @@ STATIC int ipourma_get_iflink(const struct net_device *dev)
 
 STATIC int ipourma_set_mac(struct net_device *dev, void *addr)
 {
-	// this function will be filled in the next commit
+	struct sockaddr *sa = addr;
+
+	if (IS_ERR_OR_NULL(dev) || IS_ERR_OR_NULL(addr))
+		return -EINVAL;
+	if (!is_valid_ether_addr(sa->sa_data))
+		return -EINVAL;
+	if (!(dev->priv_flags & IFF_LIVE_ADDR_CHANGE) && netif_running(dev))
+		return -EBUSY;
+
+	memcpy((void *)dev->dev_addr, sa->sa_data, dev->addr_len);
+
 	return 0;
 }
 
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_netlink.c b/drivers/ub/urma/ulp/ipourma/ipourma_netlink.c
index 9de717540a06..a42f1d2325f9 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_netlink.c
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_netlink.c
@@ -14,7 +14,99 @@
 
 #include "ipourma_netlink.h"
 
+/*
+ * fill info support: op_mode, transport mode, max_send_sge, urma_mtu
+ * The attributes info can be changed according to the actual scenario
+ */
+#define IPOURMA_ATTR_SIZE 4
+static const struct nla_policy ipourma_policy[IFLA_IPOURMA_MAX + 1] = {
+	[IFLA_IPOURMA_OP_MODE]		= { .type = NLA_U32 },
+	[IFLA_IPOURMA_TRANSPORT_MODE]	= { .type = NLA_U32 },
+	[IFLA_IPOURMA_MAX_SEND_SGE]	= { .type = NLA_U32 },
+	[IFLA_IPOURMA_XMTU]		= { .type = NLA_U32 },
+};
+
+STATIC size_t ipourma_get_size(const struct net_device *dev)
+{
+	return nla_total_size(IPOURMA_ATTR_SIZE) + /* IFLA_IPOURMA_OP_MODE */
+		nla_total_size(IPOURMA_ATTR_SIZE) + /* IFLA_IPOURMA_TRANSPORT_MODE */
+		nla_total_size(IPOURMA_ATTR_SIZE) + /* IFLA_IPOURMA_MAX_SEND_SGE */
+		nla_total_size(IPOURMA_ATTR_SIZE); /* IFLA_IPOURMA_XMTU */
+}
+
+STATIC int ipourma_fill_info(struct sk_buff *skb, const struct net_device *dev)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+
+	if (IS_ERR_OR_NULL(skb) || nla_put_u32(skb, IFLA_IPOURMA_OP_MODE, priv->urma_op_mode) ||
+		nla_put_u32(skb, IFLA_IPOURMA_TRANSPORT_MODE, priv->urma_transport_mode) ||
+		nla_put_u32(skb, IFLA_IPOURMA_MAX_SEND_SGE, priv->max_send_sge) ||
+		nla_put_u32(skb, IFLA_IPOURMA_XMTU, priv->urma_mtu)) {
+		pr_err("fill_info: put attribute failed\n");
+		return -EMSGSIZE;
+	}
+
+	return 0;
+}
+
+/*
+ * The ipourma does not support manual adding of ipourma devices through ip commands.
+ * Therefore, only the minimum device setup is performed here to comply with the
+ * RTnetlink framework.
+ */
+STATIC void ipourma_setup_common(struct net_device *dev)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+
+	priv->dev = dev;
+	dev->type = ARPHRD_ETHER;
+	dev->mtu = IPOURMA_DEFAULT_MTU;
+	dev->hard_header_len = ETH_HLEN;
+	dev->addr_len = ETH_ALEN;
+	dev->flags = IFF_NOARP;
+	dev->needs_free_netdev = false;
+}
+
+/*
+ * The ipourma does not support manual adding of ipourma devices through ip commands.
+ * The feedback info will be displayed if you enter command like
+ * 'ip link add name ipourma3 type ipourma'.
+ */
+STATIC int ipourma_newlink(struct net *src_net, struct net_device *dev,
+				struct nlattr *tb[], struct nlattr *data[],
+				struct netlink_ext_ack *extack)
+{
+	NL_SET_ERR_MSG_MOD(extack, "ipourma does not support manual device creation");
+	return -EOPNOTSUPP;
+}
+
+/*
+ * By default, the RTnetlink framework supports the modification of standard kernel
+ * attributes using the ip command.
+ * Example: ip link set dev ipourma0 up  Therefore, no additional operation in function
+ * ipourma_changelink is required.
+ * User-defined attributes need to be obtained by the user-mode client through the
+ * interface of the RTnetlink framework.
+ */
+STATIC int ipourma_changelink(struct net_device *dev, struct nlattr *tb[],
+				struct nlattr *data[], struct netlink_ext_ack *extack)
+{
+	return 0;
+}
+
+struct rtnl_link_ops ipourma_link_ops __read_mostly = {
+	.kind       = "ipourma",
+	.maxtype    = IFLA_IPOURMA_MAX,
+	.policy     = ipourma_policy,
+	.priv_size  = sizeof(struct ipourma_dev_priv),
+	.setup      = ipourma_setup_common,
+	.newlink    = ipourma_newlink,
+	.changelink = ipourma_changelink,
+	.get_size   = ipourma_get_size,
+	.fill_info  = ipourma_fill_info,
+};
+
 void ipourma_set_rtnl_link_ops(struct net_device *dev)
 {
-	// this function will be filled in the next commit
+	dev->rtnl_link_ops = &ipourma_link_ops;
 }
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_res.c b/drivers/ub/urma/ulp/ipourma/ipourma_res.c
index 491e499c649e..c5b803b23c3c 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_res.c
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_res.c
@@ -477,6 +477,113 @@ int ipourma_reset_rings(struct ipourma_dev_priv *priv)
 	return IPOURMA_OK;
 }
 
+STATIC inline void ipourma_restart_rx_segments(struct ipourma_dev_priv *priv,
+						struct ipourma_rx_buf *rx_req)
+{
+	struct ubcore_seg_cfg cfg = { 0 };
+
+	ipourma_build_seg_cfg(&cfg, (u64)rx_req->buf_aligned, priv->skb_buf_size);
+	ipourma_register_rx_segments(priv->dev, &cfg, rx_req);
+}
+
+int ipourma_restart_rings(struct ipourma_dev_priv *priv)
+{
+	struct ubcore_seg_cfg cfg = { 0 };
+
+	if (IS_ERR_OR_NULL(priv->jetty) || IS_ERR_OR_NULL(priv->jfr))
+		return IPOURMA_OK;
+	for (int i = 0; i < IPOURMA_MAX_EID_CNT; i++) {
+		if (eid_is_empty(&priv->eid_info[i].eid))
+			continue;
+		if (!IS_ERR_OR_NULL(priv->jetty[i]) || !IS_ERR_OR_NULL(priv->jfr[i]))
+			continue;
+		ipourma_init_urma_resources_by_eid(priv, i);
+		ipourma_init_tx_bufs(priv, i);
+
+		for (u32 j = 0; j < priv->rx_buf_num; j++) {
+			ipourma_build_seg_cfg(&cfg, (u64)priv->rx_buf_aligned[i][j],
+							IPOURMA_REGISTER_SEG_SIZE);
+			priv->ipourma_ub_rx_seg[i][j] = ubcore_register_seg(priv->urma_dev,
+											&cfg, NULL);
+		}
+		for (u32 j = 0; j < ipourma_rx_ring_size; j++) {
+			ipourma_restart_rx_segments(priv, &priv->rx_ring[i][j]);
+			ipourma_urma_post_recv(priv->dev, i, j);
+		}
+	}
+	return IPOURMA_OK;
+}
+
+STATIC int ipourma_init_rings_tables(struct net_device *dev)
+{
+	size_t size = sizeof(spinlock_t) * IPOURMA_MAX_EID_CNT;
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+	int cnt = IPOURMA_MAX_EID_CNT;
+
+	priv->tx_head = kcalloc(cnt, sizeof(u32), GFP_KERNEL);
+	if (IS_ERR_OR_NULL(priv->tx_head))
+		goto tx_head_failed;
+	priv->tx_tail = kcalloc(cnt, sizeof(u32), GFP_KERNEL);
+	if (IS_ERR_OR_NULL(priv->tx_tail))
+		goto tx_tail_failed;
+	priv->tx_count = kcalloc(cnt, sizeof(u32), GFP_KERNEL);
+	if (IS_ERR_OR_NULL(priv->tx_count))
+		goto tx_count_failed;
+	priv->tx_ring = kcalloc(cnt, sizeof(struct ipourma_tx_buf *), GFP_KERNEL);
+	if (IS_ERR_OR_NULL(priv->tx_ring))
+		goto tx_ring_failed;
+	priv->rx_ring = kcalloc(cnt, sizeof(struct ipourma_rx_buf *), GFP_KERNEL);
+	if (IS_ERR_OR_NULL(priv->rx_ring))
+		goto rx_ring_failed;
+	priv->tx_ring_locks = kzalloc(size, GFP_KERNEL);
+	if (IS_ERR_OR_NULL(priv->tx_ring_locks))
+		goto tx_ring_locks_failed;
+	return IPOURMA_OK;
+
+tx_ring_locks_failed:
+	kfree(priv->rx_ring);
+rx_ring_failed:
+	kfree(priv->tx_ring);
+tx_ring_failed:
+	kfree(priv->tx_count);
+tx_count_failed:
+	kfree(priv->tx_tail);
+tx_tail_failed:
+	kfree(priv->tx_head);
+tx_head_failed:
+	netdev_err(priv->dev, "%s\n", ipourma_err_desc(IPOURMA_INIT_RINGS_TABLE_FAILED));
+	return IPOURMA_INIT_RINGS_TABLE_FAILED;
+}
+
+/**
+ * @note: tx_buf & rx_buf are initialized dynamically
+ */
+int ipourma_init_rings(struct net_device *dev)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+	int ret = IPOURMA_OK;
+
+	priv->skb_buf_size = priv->urma_mtu < IPOURMA_SEGMENT_ALIGN_SIZE ?
+					IPOURMA_SEGMENT_ALIGN_SIZE : priv->urma_mtu;
+
+	ret = ipourma_init_rings_tables(dev);
+	if (ret != IPOURMA_OK)
+		return ret;
+
+	if (priv->urma_mtu > IPOURMA_SEGMENT_ALIGN_SIZE) {
+		priv->tx_buf_size = (priv->urma_mtu + IPOURMA_SEGMENT_ALIGN_SIZE - 1) &
+							~(IPOURMA_SEGMENT_ALIGN_SIZE - 1);
+	} else {
+		priv->tx_buf_size = IPOURMA_SEGMENT_ALIGN_SIZE;
+	}
+	priv->rx_buf_num = DIV_ROUND_UP(ipourma_rx_ring_size * priv->skb_buf_size,
+					IPOURMA_REGISTER_SEG_SIZE);
+	priv->tx_buf_num = DIV_ROUND_UP(ipourma_tx_ring_size * priv->tx_buf_size,
+					IPOURMA_REGISTER_SEG_SIZE);
+
+	return ret;
+}
+
 void ipourma_uninit_tjetty_hmap(struct net_device *dev)
 {
 	struct ipourma_dev_priv *priv = netdev_priv(dev);
@@ -560,6 +667,14 @@ STATIC struct ubcore_jetty *ipourma_create_jetty(struct net_device *dev,
 	return ubcore_create_jetty(priv->urma_dev, &jetty_cfg, NULL, NULL);
 }
 
+STATIC struct ubcore_jfc *ipourma_create_jfc(struct net_device *dev,
+					     ubcore_comp_callback_t jfce_handler,
+					     u32 depth)
+{
+	// this function will be filled in the next commit
+	return NULL;
+}
+
 STATIC void ipourma_uninit_misc(struct ipourma_dev_priv *priv)
 {
 	if (!IS_ERR_OR_NULL(priv->tjetty_lru.tjetty_wq)) {
@@ -629,6 +744,67 @@ void ipourma_uninit_urma_resources(struct net_device *dev)
 	}
 }
 
+STATIC int ipourma_init_misc(struct ipourma_dev_priv *priv)
+{
+	priv->max_send_sge = IPOURMA_MAX_URMA_SEND_SGES;
+	priv->urma_mtu = IPOURMA_URMA_MAX_MTU;
+	priv->urma_op_mode = UBCORE_OPC_SEND;
+	priv->urma_transport_mode = UBCORE_TP_RM;
+
+	priv->tjetty_lru.tjetty_wq = alloc_workqueue("ipourma_tjetty_wq", WQ_MEM_RECLAIM, 0);
+	if (IS_ERR_OR_NULL(priv->tjetty_lru.tjetty_wq))
+		goto tjetty_wq_failed;
+
+	priv->tx_wq = alloc_ordered_workqueue("ipourma_tx_wq", 0);
+	if (IS_ERR_OR_NULL(priv->tx_wq))
+		goto tx_wq_failed;
+
+	priv->rx_wq = alloc_workqueue("ipourma_rx_wq", WQ_MEM_RECLAIM, 0);
+	if (IS_ERR_OR_NULL(priv->rx_wq))
+		goto rx_wq_failed;
+
+	/* Net configurations should be called in order. */
+	priv->net_config_wq = alloc_ordered_workqueue("net_config_wq", 0);
+	if (IS_ERR_OR_NULL(priv->net_config_wq))
+		goto net_config_wq_failed;
+
+	return IPOURMA_OK;
+net_config_wq_failed:
+	destroy_workqueue(priv->rx_wq);
+	priv->rx_wq = NULL;
+rx_wq_failed:
+	destroy_workqueue(priv->tx_wq);
+	priv->tx_wq = NULL;
+tx_wq_failed:
+	destroy_workqueue(priv->tjetty_lru.tjetty_wq);
+	priv->tjetty_lru.tjetty_wq = NULL;
+tjetty_wq_failed:
+	return -1;
+}
+
+STATIC int ipourma_init_urma_resources_table(struct net_device *dev)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+	int ret = IPOURMA_OK;
+
+	priv->jfr = kcalloc(IPOURMA_MAX_EID_CNT, sizeof(struct ubcore_jfr *), GFP_KERNEL);
+	if (IS_ERR_OR_NULL(priv->jfr)) {
+		ret = IPOURMA_CREATE_JFR_TABLE_FAILED;
+		goto jfr_table_failed;
+	}
+	priv->jetty = kcalloc(IPOURMA_MAX_EID_CNT, sizeof(struct ubcore_jetty *), GFP_KERNEL);
+	if (IS_ERR_OR_NULL(priv->jetty)) {
+		ret = IPOURMA_CREATE_JETTY_TABLE_FAILED;
+		goto jetty_table_failed;
+	}
+	return ret;
+jetty_table_failed:
+	kfree(priv->jfr);
+	priv->jfr = NULL;
+jfr_table_failed:
+	return ret;
+}
+
 static bool ipourma_check_dev_name(struct net_device *dev)
 {
 	char buf[IPOURMA_MAX_DEV_NAME] = {0};
@@ -678,3 +854,51 @@ int ipourma_init_urma_resources_by_eid(struct ipourma_dev_priv *priv, u32 eid_id
 invalid_name:
 	return ret;
 }
+
+int ipourma_init_urma_resources(struct net_device *dev)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+	int ret = IPOURMA_OK;
+
+	if (!ipourma_check_dev_name(dev))
+		goto invalid_name;
+
+	ret = ipourma_init_urma_resources_table(dev);
+	if (ret != IPOURMA_OK)
+		goto init_table_failed;
+
+	priv->tx_jfc = ipourma_create_jfc(dev, ipourma_handle_tx_cqe,
+								IPOURMA_TX_JFC_DEPTH);
+	if (IS_ERR_OR_NULL(priv->tx_jfc)) {
+		ret = IPOURMA_CREATE_JFC_FAILED;
+		pr_err("create tx jfc error, dev: %s\n", dev->name);
+		goto tx_jfc_failed;
+	}
+	priv->rx_jfc = ipourma_create_jfc(dev, ipourma_handle_rx_cqe,
+								IPOURMA_RX_JFC_DEPTH);
+	if (IS_ERR_OR_NULL(priv->rx_jfc)) {
+		ret = IPOURMA_CREATE_JFC_FAILED;
+		pr_err("create rx jfc error, dev: %s\n", dev->name);
+		goto rx_jfc_failed;
+	}
+
+	ret = ipourma_init_misc(priv);
+	if (ret != IPOURMA_OK) {
+		pr_err("%s create wq failed.\n", dev->name);
+		goto init_misc_failed;
+	}
+
+	return ret;
+
+init_misc_failed:
+	ubcore_delete_jfc(priv->rx_jfc);
+rx_jfc_failed:
+	ubcore_delete_jfc(priv->tx_jfc);
+	priv->tx_jfc = NULL;
+tx_jfc_failed:
+	ipourma_uninit_urma_resources_table(priv->dev);
+init_table_failed:
+	return ret;
+invalid_name:
+	return ret;
+}
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_res.h b/drivers/ub/urma/ulp/ipourma/ipourma_res.h
index 25fe4e72bc73..439691283a28 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_res.h
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_res.h
@@ -12,6 +12,8 @@
 
 void ipourma_uninit_rings_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx);
 void ipourma_uninit_rx_bufs(struct ipourma_dev_priv *priv, u32 eid_idx);
+int ipourma_restart_rings(struct ipourma_dev_priv *priv);
+int ipourma_init_rings(struct net_device *dev);
 void ipourma_uninit_rings(struct net_device *dev);
 int ipourma_init_tjetty_hmap(struct net_device *dev);
 void ipourma_uninit_tjetty_hmap(struct net_device *dev);
@@ -20,5 +22,6 @@ int ipourma_reset_rings(struct ipourma_dev_priv *priv);
 int ipourma_init_rings_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx);
 void ipourma_uninit_urma_resources_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx);
 int ipourma_init_urma_resources_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx);
+int ipourma_init_urma_resources(struct net_device *dev);
 
 #endif
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_types.h b/drivers/ub/urma/ulp/ipourma/ipourma_types.h
index 38138e6e9b09..1f2449cc628d 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_types.h
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_types.h
@@ -69,7 +69,6 @@ enum {
 					sizeof(struct ipourma_header)),
 	IPOURMA_MIN_MTU             = 1280,
 	IPOURMA_DEFAULT_MTU         = IPOURMA_MAX_MTU,
-	IPOURMA_MIN_SKB_LEN         = 68,
 	IPOURMA_ALEN                = 6,
 	IPOURMA_DEFAULT_TJETTY_CAP  = 256,
 	IPOURMA_MAX_EID_CNT         = 32,
@@ -261,6 +260,8 @@ struct ipourma_tjetty_hash_node {
 	struct list_head lru_list;
 	union ubcore_eid key[2];
 	struct ubcore_tjetty *tjetty;
+	struct work_struct unimport_work;
+	struct ipourma_tjetty_lru *tjetty_lru;
 	u64 last_jiffies;
 };
 
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_ub.c b/drivers/ub/urma/ulp/ipourma/ipourma_ub.c
index 496fa701c1f7..476cf04c0fa4 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_ub.c
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_ub.c
@@ -91,14 +91,58 @@ STATIC int ipourma_setup_sges(struct ipourma_dev_priv *priv,
 	return (nr_frags + offset);
 }
 
+STATIC inline uint32_t hash_eids(union ubcore_eid *src_eid,
+				 union ubcore_eid *dst_eid,
+				 uint32_t hash_seed)
+{
+	// this function will be filled in the next commit
+	return 0;
+}
+
+STATIC struct ubcore_tjetty *ipourma_import_jetty(struct net_device *dev,
+	union ubcore_eid *dst_eid, uint32_t jetty_id)
+{
+	// this function will be filled in the next commit
+	return NULL;
+}
+
+STATIC inline void ipourma_renew_tjetty_node(struct ipourma_tjetty_lru *tjetty_lru,
+					  struct ipourma_tjetty_hash_node *tjetty_node)
+{
+	// this function will be filled in the next commit
+}
+
 STATIC struct ipourma_tjetty_hash_node *ipourma_locate_tjetty_node(
 					struct ipourma_tjetty_lru *tjetty_lru,
 					union ubcore_eid *src_eid,
 					union ubcore_eid *dst_eid,
 					bool lock_free)
+{
+	struct ipourma_tjetty_hmap *tjetty_hmap = &tjetty_lru->tjetty_hmap;
+	uint32_t hash_key = hash_eids(src_eid, dst_eid, tjetty_hmap->hash_seed);
+	struct ipourma_tjetty_hash_node *tjetty_node = NULL;
+
+	hash_key = hash_key & (IPOURMA_TJETTY_HMAP_SIZE - 1);
+	if (!lock_free)
+		spin_lock(&tjetty_lru->lock);
+	hlist_for_each_entry(tjetty_node, &tjetty_hmap->buckets[hash_key], hlist) {
+		if (ipourma_are_eids_equal(&tjetty_node->key[0], src_eid) &&
+		    ipourma_are_eids_equal(&tjetty_node->key[1], dst_eid)) {
+			pr_debug("dst_eid hit in hmap\n");
+			ipourma_renew_tjetty_node(tjetty_lru, tjetty_node);
+			break;
+		}
+	}
+	if (!lock_free)
+		spin_unlock(&tjetty_lru->lock);
+
+	return tjetty_node;
+}
+
+STATIC void ipourma_insert_tjetty_node(struct ipourma_tjetty_lru *tjetty_lru,
+					struct ipourma_tjetty_hash_node *tjetty_node)
 {
 	// this function will be filled in the next commit
-	return NULL;
 }
 
 STATIC inline void ipourma_delete_tjetty_node(struct ipourma_tjetty_hash_node *tjetty_node)
@@ -136,6 +180,12 @@ void ipourma_lru_clear(struct ipourma_tjetty_lru *tjetty_lru)
 	spin_unlock(&tjetty_lru->lock);
 }
 
+STATIC void ipourma_lru_update(struct ipourma_tjetty_lru *tjetty_lru,
+				struct ipourma_tjetty_hash_node *tjetty_node)
+{
+	// this function will be filled in the next commit
+}
+
 STATIC void tjetty_aging_callback(struct work_struct *work)
 {
 	struct delayed_work *tjetty_aging_work = container_of(work,
@@ -183,10 +233,45 @@ void ipourma_init_tjetty_aging_work(struct ipourma_tjetty_lru *tjetty_lru)
 	spin_unlock_irqrestore(&tjetty_lru->lock, flags);
 }
 
+STATIC void ipourma_unimport_tjetty_cb(struct work_struct *work)
+{
+	// this function will be filled in the next commit
+}
+
 STATIC struct ubcore_tjetty *ipourma_import_new_tjetty(
 	struct ipourma_dev_priv *priv, struct ipourma_tx_buf *tx_req)
 {
-	// this function will be filled in the next commit
+	if (IS_ERR_OR_NULL(priv) || IS_ERR_OR_NULL(tx_req))
+		goto nullptr_err;
+	u32 eid_index = tx_req->eid_index;
+	union ubcore_eid src_eid = priv->eid_info[eid_index].eid;
+	union ubcore_eid dst_eid = tx_req->dst_eid;
+	struct ipourma_tjetty_hash_node *tjetty_node = NULL;
+	u32 jetty_id = eid_index + IPOURMA_WELL_KNOWN_JETTY_ID;
+
+	tjetty_node = kzalloc(sizeof(struct ipourma_tjetty_hash_node), GFP_KERNEL);
+	if (IS_ERR_OR_NULL(tjetty_node)) {
+		netdev_dbg(priv->dev, "%s\n",
+				ipourma_err_desc(IPOURMA_TJETTY_NODE_ALLOC_FAILED));
+		goto tjetty_node_zalloc_failed;
+	}
+
+	tjetty_node->key[0] = src_eid;
+	tjetty_node->key[1] = dst_eid;
+	tjetty_node->tjetty_lru = &priv->tjetty_lru;
+	INIT_WORK(&tjetty_node->unimport_work, ipourma_unimport_tjetty_cb);
+	tjetty_node->tjetty = ipourma_import_jetty(priv->dev, &dst_eid, jetty_id);
+	if (IS_ERR_OR_NULL(tjetty_node->tjetty))
+		goto ipourma_import_jetty_failed;
+
+	ipourma_insert_tjetty_node(&priv->tjetty_lru, tjetty_node);
+	ipourma_lru_update(&priv->tjetty_lru, tjetty_node);
+	return tjetty_node->tjetty;
+nullptr_err:
+	return NULL;
+ipourma_import_jetty_failed:
+	kfree(tjetty_node);
+tjetty_node_zalloc_failed:
 	return NULL;
 }
 
@@ -289,6 +374,63 @@ void ipourma_post_send(struct work_struct *work)
 	tx_req->skb = NULL;
 }
 
+STATIC inline int ipourma_get_eid_index(struct net_device *dev, union ubcore_eid *eid)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+
+	for (u32 i = 0; i < IPOURMA_MAX_EID_CNT; i++) {
+		if (!eid_is_empty(&priv->eid_info[i].eid)
+				&& ipourma_are_eids_equal(&priv->eid_info[i].eid, eid))
+			return i;
+	}
+	return -IPOURMA_SRC_IP_ADDR_EID_MISMATCH;
+}
+
+int ipourma_check_skb(struct net_device *dev, struct sk_buff *skb)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+
+	if (unlikely(skb_is_gso(skb) == true)) {
+		/**
+		 *  upper layer expects ipourma to do the GSO,
+		 *  while ipourma or urma device doesn't support GSO
+		 */
+		priv->runtime_stats.tx_stats.gso_not_support++;
+		netdev_dbg(dev, "%s: gso_size = %d\n",
+					ipourma_err_desc(IPOURMA_NOT_SUPPORT_GSO),
+					skb_shinfo(skb)->gso_size);
+		return IPOURMA_NOT_SUPPORT_GSO;
+	}
+	if (unlikely(skb->len > priv->urma_mtu)) {
+		/**
+		 *  the overall length of linear & nonlinear should not
+		 *  exceed the urma mtu when using send operation over CTP
+		 *  ubcore doesn't define a CTP mode, use UBCORE_TP_RM for now
+		 */
+		if ((priv->urma_op_mode == UBCORE_OPC_SEND) &&
+			(priv->urma_transport_mode == UBCORE_TP_RM)) {
+			priv->runtime_stats.tx_stats.packet_size_error++;
+			netdev_dbg(dev, "%s: skb len = %u\n",
+						ipourma_err_desc(IPOURMA_GIANT_PACKET),
+						skb->len);
+			return IPOURMA_GIANT_PACKET;
+		}
+	}
+	if (unlikely(skb_shinfo(skb)->nr_frags + 1 > priv->max_send_sge)) {
+		/* try to linearize firstly */
+		skb_linearize(skb);
+		if (skb_shinfo(skb)->nr_frags + 1 > priv->max_send_sge) {
+			priv->runtime_stats.tx_stats.frag_error++;
+			netdev_dbg(dev, "%s: skb nr_frags = %d, max_send_sge = %d\n",
+						ipourma_err_desc(IPOURMA_TOO_MANY_FRAGS),
+						skb_shinfo(skb)->nr_frags, priv->max_send_sge);
+			return IPOURMA_TOO_MANY_FRAGS;
+		}
+	}
+
+	return IPOURMA_OK;
+}
+
 int ipourma_register_rx_segments(struct net_device *dev,
 	struct ubcore_seg_cfg *cfg, struct ipourma_rx_buf *rx_req)
 {
@@ -312,6 +454,87 @@ int ipourma_register_rx_segments(struct net_device *dev,
 	return IPOURMA_OK;
 }
 
+STATIC inline int ipourma_add_header(struct sk_buff *skb)
+{
+	struct ipourma_header *header;
+	size_t header_size = sizeof(struct ipourma_header);
+
+	/* For now, addresses are not used */
+	header = skb_push(skb, header_size);
+	header->proto = skb->protocol;
+
+	return header_size;
+}
+
+int ipourma_xmit(struct net_device *dev, struct sk_buff *skb,
+		 union ubcore_eid *src_eid, union ubcore_eid *dst_eid)
+{
+	struct ipourma_tjetty_hash_node *tjetty_node = NULL;
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+	struct ipourma_tx_buf *tx_req;
+	int ret = IPOURMA_OK;
+	unsigned long flags;
+	u32 eid_idx;
+	u32 tx_idx;
+
+	ipourma_add_header(skb);
+	ret = ipourma_get_eid_index(dev, src_eid);
+	if (unlikely(ret == -IPOURMA_SRC_IP_ADDR_EID_MISMATCH)) {
+		priv->runtime_stats.tx_stats.ip_eid_not_equal++;
+		netdev_dbg(dev, "%s:IP=0x%llx-0x%llx\n",
+			ipourma_err_desc(IPOURMA_SRC_IP_ADDR_EID_MISMATCH),
+			src_eid->in6.subnet_prefix,
+			src_eid->in6.interface_id);
+		return IPOURMA_SRC_IP_ADDR_EID_MISMATCH;
+	}
+	eid_idx = (u32)ret;
+	ret = IPOURMA_OK;
+	if (spin_trylock(&priv->tjetty_lru.lock)) {
+		tjetty_node = ipourma_locate_tjetty_node(&priv->tjetty_lru, src_eid, dst_eid, true);
+		spin_unlock(&priv->tjetty_lru.lock);
+	}
+
+	// enqueue the skb
+	spin_lock_irqsave(&priv->tx_ring_locks[eid_idx], flags);
+	if (priv->tx_ring_is_full[eid_idx]) {
+		spin_unlock_irqrestore(&priv->tx_ring_locks[eid_idx], flags);
+		priv->runtime_stats.tx_stats.tx_ring_full++;
+		return IPOURMA_TX_RING_FULL;
+	}
+	tx_idx = priv->tx_head[eid_idx] & (ipourma_tx_ring_size - 1);
+	tx_req = &priv->tx_ring[eid_idx][tx_idx];
+	tx_req->tx_buf_in_use = 1;
+	if ((priv->tx_head[eid_idx] - priv->tx_tail[eid_idx]) == ipourma_tx_ring_size - 1) {
+		/* tx ring is full, notify the upper layer */
+		priv->tx_ring_is_full[eid_idx] = true;
+		atomic_add(1, &priv->tx_ring_blocked);
+		netif_stop_queue(dev);
+		netdev_dbg(dev, "%s: head = %u, tail = %u\n",
+					ipourma_err_desc(IPOURMA_TX_RING_FULL),
+					priv->tx_head[eid_idx], priv->tx_tail[eid_idx]);
+	}
+	priv->tx_head[eid_idx]++;
+	spin_unlock_irqrestore(&priv->tx_ring_locks[eid_idx], flags);
+	tx_req->skb = skb;
+	tx_req->dst_eid = *dst_eid;
+	tx_req->eid_index = eid_idx;
+	tx_req->tjetty = IS_ERR_OR_NULL(tjetty_node) ? NULL : tjetty_node->tjetty;
+
+	/* take the ownership of skb */
+	skb_orphan(skb);
+	skb_dst_drop(skb);
+	pr_debug("queue work, idx %u\n", tx_idx);
+	if (!IS_ERR_OR_NULL(tx_req->tjetty)) {
+		priv->runtime_stats.tx_stats.post_send_bypass++;
+		ipourma_post_send(&(tx_req->work));
+	} else {
+		priv->runtime_stats.tx_stats.post_send_enque++;
+		queue_work(priv->tx_wq, &(tx_req->work));
+	}
+
+	return ret;
+}
+
 STATIC struct sk_buff *ipourma_alloc_rx_skb(struct net_device *dev)
 {
 	struct ipourma_dev_priv *priv = netdev_priv(dev);
@@ -424,6 +647,28 @@ void ipourma_replenish_segments(struct work_struct *work)
 		ipourma_urma_post_recv(priv->dev, rx_buf->eid_index, rx_buf->idx);
 }
 
+int ipourma_napi_tx_poll(struct napi_struct *napi, int budget)
+{
+	// this function will be filled in the next commit
+	return 0;
+}
+
+int ipourma_napi_rx_poll(struct napi_struct *napi, int budget)
+{
+	// this function will be filled in the next commit
+	return 0;
+}
+
+STATIC void ipourma_napi_add(struct net_device *dev)
+{
+	struct ipourma_dev_priv *priv = netdev_priv(dev);
+
+	netif_napi_add_weight(dev, &priv->napi_send, ipourma_napi_tx_poll,
+						  IPOURMA_NAPI_TX_WEIGHT);
+	netif_napi_add_weight(dev, &priv->napi_recv, ipourma_napi_rx_poll,
+						  IPOURMA_NAPI_RX_WEIGHT);
+}
+
 STATIC inline void ipourma_napi_del(struct net_device *dev)
 {
 	struct ipourma_dev_priv *priv = netdev_priv(dev);
@@ -457,6 +702,37 @@ int ipourma_urma_init_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx)
 	return ret;
 }
 
+void ipourma_handle_rx_cqe(struct ubcore_jfc *jfc)
+{
+	// this function will be filled in the next commit
+}
+
+void ipourma_handle_tx_cqe(struct ubcore_jfc *jfc)
+{
+	// this function will be filled in the next commit
+}
+
+int ipourma_urma_dev_init(struct net_device *dev)
+{
+	int ret;
+
+	ipourma_napi_add(dev);
+	ret = ipourma_init_rings(dev);
+	if (ret != IPOURMA_OK)
+		goto init_rings_failed;
+
+	ret = ipourma_init_urma_resources(dev);
+	if (ret != IPOURMA_OK)
+		goto init_urma_res_failed;
+
+	return ret;
+init_urma_res_failed:
+	ipourma_uninit_rings(dev);
+init_rings_failed:
+	ipourma_napi_del(dev);
+	return ret;
+}
+
 void ipourma_urma_dev_uninit(struct net_device *dev)
 {
 	ipourma_uninit_tjetty_hmap(dev);
diff --git a/drivers/ub/urma/ulp/ipourma/ipourma_ub.h b/drivers/ub/urma/ulp/ipourma/ipourma_ub.h
index bfc15632e832..5fab230f1192 100644
--- a/drivers/ub/urma/ulp/ipourma/ipourma_ub.h
+++ b/drivers/ub/urma/ulp/ipourma/ipourma_ub.h
@@ -15,8 +15,16 @@ void ipourma_build_seg_cfg(struct ubcore_seg_cfg *cfg,
 void ipourma_lru_clear(struct ipourma_tjetty_lru *tjetty_lru);
 void ipourma_init_tjetty_aging_work(struct ipourma_tjetty_lru *tjetty_lru);
 int ipourma_urma_init_by_eid(struct ipourma_dev_priv *priv, u32 eid_idx);
+void ipourma_handle_rx_cqe(struct ubcore_jfc *jfc);
+void ipourma_handle_tx_cqe(struct ubcore_jfc *jfc);
+int ipourma_urma_dev_init(struct net_device *dev);
 void ipourma_urma_dev_uninit(struct net_device *dev);
 void ipourma_post_send(struct work_struct *work);
+int ipourma_xmit(struct net_device *dev, struct sk_buff *skb,
+		 union ubcore_eid *src_eid, union ubcore_eid *dst_eid);
+int ipourma_check_skb(struct net_device *dev, struct sk_buff *skb);
+int ipourma_register_rx_segments(struct net_device *dev,
+	struct ubcore_seg_cfg *cfg, struct ipourma_rx_buf *rx_req);
 int ipourma_urma_post_recv(struct net_device *dev, u32 eid_idx, u32 idx);
 void ipourma_replenish_segments(struct work_struct *work);
 void ipourma_handle_tx_wc(struct net_device *dev,
-- 
2.43.0

